hypothesis-python/src/hypothesis/internal/constants_ast.py

```diff
@@ -0,0 +1,136 @@
+
+ConstantT: "TypeAlias" = Union[int, float, bool, bytes, str]
+
+
+class ConstantVisitor(NodeVisitor):
+    def __init__(self):
+        super().__init__()
+        self.constants: set[ConstantT] = set()
+
+    def _add_constant(self, constant: object) -> None:
+        self.constants |= self._unfold_constant(constant)
+
+    def visit_UnaryOp(self, node: UnaryOp) -> None:
+        # `a = -1` is actually a combination of a USub and the constant 1.
+        if (
+            isinstance(node.op, USub)
+            and isinstance(node.operand, Constant)
+            and isinstance(node.operand.value, (int, float))
+            and not isinstance(node.operand.value, bool)
+        ):
+            self._add_constant(-node.operand.value)
+            # don't recurse on this node to avoid adding the positive variant
+            return
+
+        self.generic_visit(node)
+
+    def visit_Expr(self, node: Expr) -> None:
+        if isinstance(node.value, Constant) and isinstance(node.value.value, str):
+            return
+
+        self.generic_visit(node)
+
+    def visit_JoinedStr(self, node):
+        # dont recurse on JoinedStr, i.e. f strings. Constants that appear *only*
+        # in f strings are unlikely to be helpful.
+        return
+
+    @classmethod
+    def _unfold_constant(cls, value: object) -> set[ConstantT]:
+        if isinstance(value, str) and (
+            len(value) > 20 or value.isspace() or value == ""
+        ):
+            # discard long strings, which are unlikely to be useful.
+            return set()
+        if isinstance(value, bool):
+            return set()
+        if isinstance(value, float) and math.isinf(value):
+            # we already upweight inf.
+            return set()
+        if isinstance(value, (int, float, bytes, str)):
+            return {value}
+        # I don't kow what case could go here, but am also not confident there
+        # isn't one.
+        return set()  # pragma: no cover
+
+    def visit_Constant(self, node):
+        self._add_constant(node.value)
+        self.generic_visit(node)
+
+
+@lru_cache(1024)
+def constants_from_ast(tree: AST) -> set[ConstantT]:
+    visitor = ConstantVisitor()
+    visitor.visit(tree)
+    return visitor.constants
+
+
+@lru_cache(1024)
+def _module_ast(module: ModuleType) -> Optional[AST]:
+    try:
+        source = inspect.getsource(module)
+        tree = ast.parse(source)
+    except Exception:
+        return None
+
+    return tree
+
+
+def local_modules() -> tuple[ModuleType, ...]:
+    modules = []
+    for module in sys.modules.values():
+        if not hasattr(module, "__file__"):
+            continue
+        if module.__file__ is None:  # pragma: no cover
+            continue
+
+        if ModuleLocation.from_path(module.__file__) is not ModuleLocation.LOCAL:
+            continue
+
+        modules.append(module)
+    return tuple(modules)
+
+
+def local_constants():
+    constants = set()
+    for module in local_modules():
```

Comment:
for later: I think we'll want to make this cache both the modules it's seen before (by name), and the derived constants; that way we only pay the processing cost for newly-imported local modules on re-runs.